File: app/src/main/java/ws/grigory/currencycalculator/calculator/MainDisplay.java
package ws.grigory.currencycalculator.calculator;

import static ws.grigory.currencycalculator.CCWidgetLarge.BS;
import static ws.grigory.currencycalculator.CCWidgetLarge.DF;
import static ws.grigory.currencycalculator.CCWidgetLarge.DS;
import static ws.grigory.currencycalculator.CCWidgetLarge.EVAL;
import static ws.grigory.currencycalculator.CCWidgetLarge.ZERO_CHAR;
import static ws.grigory.currencycalculator.WidgetParameters.checkInfinity;

import java.text.ParseException;

import java.util.ArrayList;
import java.util.Objects;

import ws.grigory.currencycalculator.Currency;
public class MainDisplay extends Display {

    public final StringBuilder expression = new StringBuilder();

    public MainDisplay(int currencyIndex, ArrayList<Currency> currencies) {
        super(currencyIndex, currencies);
    }

    public Currency getMainDisplayCurrency() {
        return currencies.get(currencyIndex);
    }

    public boolean processDisplayChar(boolean isNewValue, char c) {
        if (isNewValue) {
            clearDisplay();
        }

        int displayLength = display.length();
        if (BS == c) {
            if (displayLength > 1) {
                display.delete(displayLength - 1, displayLength);
                isNewValue = false;
            } else {
                clearDisplay();
                display.append(ZERO_CHAR);
                isNewValue = true;
            }
        } else {

            int indexDS = display.indexOf(Character.toString(DS));

            if (indexDS < 0 || (c != DS && displayLength - indexDS < 3)) {
                if (displayLength == 1 && display.charAt(0) == ZERO_CHAR && c != DS) {
                    display.delete(0, 1);
                }
                if(displayLength == 0 && c == DS){
                    display.append(ZERO_CHAR);
                }
                display.append(c);
                isNewValue = false;
            }
        }
        return isNewValue;
    }

    @Override
    public void writeCurrencyValueToDisplay() {
        clearDisplay();
        display.append(floatToString(currencies.get(currencyIndex).value));
    }

    public void writeDisplayValueToCurrency() {
        try {
            getMainDisplayCurrency().value = checkInfinity(Objects.requireNonNull(
                    DF.parse(display.toString())).floatValue());

            if(getMainDisplayCurrency().isInfinity()){
                writeCurrencyValueToDisplay();
            }

        } catch (ParseException ignored) {
        }
    }

    public void createExpression(char operation, float yRegistry) {
        expression.setLength(0);
        if (!getMainDisplayCurrency().isInfinity()) {
            if (operation != EVAL) {
                expression.append(floatToString(yRegistry));
                expression.append(' ');
                expression.append(operation);
            }
        }
    }
}

File: app/src/main/java/ws/grigory/currencycalculator/calculator/MainDisplay.kt
package ws.grigory.currencycalculator.calculator

import ws.grigory.currencycalculator.Constants.BS
import ws.grigory.currencycalculator.Constants.DS
import ws.grigory.currencycalculator.Constants.EVAL
import ws.grigory.currencycalculator.Constants.ZERO_CHAR
import ws.grigory.currencycalculator.Currency
import kotlin.collections.ArrayList

class MainDisplay(currencyIndex: Int, currencies: ArrayList<Currency>) :
    Display(currencyIndex, currencies) {
    val expression = StringBuilder()
    val mainDisplayCurrency: Currency
        get() = currencies[currencyIndex]

    fun processDisplayChar(isNewValue: Boolean, c: Char): Boolean {
        if (isNewValue) {
            clearDisplay()
        }
        var result = isNewValue
        with(display) {

            if (BS == c) {
                if (isNotEmpty()) {
                    delete(length - 1, length)
                }
                result = if (isNotEmpty())
                    false
                else {
                    clearDisplay()
                    append(ZERO_CHAR)
                    true
                }
            } else {

                val indexDS: Int = indexOf(DS)

                if (indexDS < 0 || (c != DS && length - indexDS < 3)) {
                    if (length == 1 && this[0] == ZERO_CHAR && c != DS) {
                        delete(0, 1)
                    }
                    if (isEmpty() && c == DS) {
                        append(ZERO_CHAR)
                    }
                    append(c)
                    result = false
                }
            }
        }
        return result
    }

    override fun writeCurrencyValueToDisplay() {
        clearDisplay()
        display.append(floatToString(currencies[currencyIndex].value))
    }

    fun writeDisplayValueToCurrency() {
        mainDisplayCurrency.setValue(display)

        if (mainDisplayCurrency.isInfinity()) {
            writeCurrencyValueToDisplay()
        }
    }

    fun createExpression(operation: Char, yRegistry: Float) {
        with(expression) {
            setLength(0)
            if (!mainDisplayCurrency.isInfinity() && operation != EVAL) {
                append(floatToString(yRegistry))
                append(' ')
                append(operation)
            }
        }
    }
}